Fix bug in passing parameters to the python action

ioana-blue opened the pull request:
Fixed bug and added test to chec a couple of jsons, one with null

Subsumes #906

ioana-blue commented:
This is ready for review & merge.

rabbah commented:
I'm not sure that we need a complex test for null - here's a suggested alternative:

https://github.com/openwhisk/openwhisk/compare/master...rabbah:pr-906?expand=1

rabbah commented on core/pythonAction/pythonaction.py:
```diff
@@ -61,7 +61,7 @@ def run():
     result = None
     try:
         exec(flask.g, namespace)
```
why isn't this part of init: `exec(flask.g, namespace)` - wouldn't that save the eval overhead on every run?

ioana-blue replied:
I think this evolved. At some point there was an eval part of init. I'll fix it part of this PR.

ioana-blue replied:
Actually, I think this statement in particular is not needed (subsumed by the following exec), which is really a bug and not only an optimization issue. Good catch!

ioana-blue commented:
One of your links above doesn't work for me, but I get what you're saying. Fixing.

ioana-blue commented:
@rr I addressed your comments and fixes are in place. Please take a careful look since I fixed another issue with providing feedback during init for when compilation fails (I think there was an issue opened for this, maybe from Nick). moving this to review.

rabbah commented on tests/src/actionContainers/PythonActionContainerTests.scala:
```diff
@@ -73,6 +72,30 @@ class PythonActionContainerTests extends FlatSpec
+
+            for (args <- argss) {
+                val (runCode, out) = c.run(runPayload(args))
+                println(out)
```
remove println.

rabbah commented on core/pythonAction/pythonaction.py:
```diff
@@ -35,8 +35,16 @@ def init():
+            flask.g = None
+            traceback.print_exc(file = sys.stderr)
+            sys.stderr.flush()
+            response = flask.jsonify({"error": "the action failed to compile. See logs for details." })
```
the -> The (since you capitalized elsewhere).

rabbah commented on core/pythonAction/pythonaction.py:
```diff
@@ -56,12 +64,18 @@ def run():
 
+    if flask.g is None:
+        # no code to execute
+        response = flask.jsonify({"error": "No code to execute (init failed). See logs for details." })
```
init -> compilation? since you called it compilation above.

rabbah replied:
this should never happen btw (ie if you're called when init failed, there's an internal system error).

rabbah commented:
Can you link to the issue mentioned above (in the commit and this pr).
PG?

rabbah commented:
Not sure if you meant this one: https://github.com/openwhisk/openwhisk/issues/493...?

rabbah commented on core/pythonAction/pythonaction.py:
```diff
@@ -73,7 +87,7 @@ def run():
         return response
     else:
-        response = flask.jsonify({ "error": "the action did not return a dictionary", "action_output": result })
+        response = flask.jsonify({"error": "The action did not return a dictionary", "action_output": result })
```
missing period.
